lib/message-parser.js

```diff
@@ -139,55 +226,45 @@ class MessageParser {
       payload = Buffer.from(options.data);
     }
 
-    // Generate prefix (including command and length bytes)
-    const prefix = Buffer.from('000055aa00000000000000' +
-                               (options.commandByte < 16 ? '0' : '') +
-                                options.commandByte.toString(16), 'hex');
+    // Ensure commandByte is a Number
+    if (typeof options.commandByte === 'string') {
+      options.commandByte = parseInt(options.commandByte, 16);
+    }
 
-    // Suffix is static
-    const suffix = Buffer.from('0000aa55', 'hex');
+    // Allocate buffer with room for payload + 24 bytes for
+    // prefix, sequence, command, length, crc, and suffix
+    const buffer = Buffer.alloc(payload.length + 24);
 
-    // As devices don't seem to care,
-    // just use an empty CRC for now.
-    const crc32Buffer = Buffer.from('00000000', 'hex');
+    // Add prefix, command, and length
+    buffer.writeUInt32BE(0x000055AA, 0);
+    buffer.writeUInt32BE(options.commandByte, 8);
+    buffer.writeUInt32BE(payload.length + 8, 12);
 
-    // Calculate length (everything past length byte)
-    const len = Buffer.allocUnsafe(4);
-    len.writeInt32BE(Buffer.concat([payload, crc32Buffer, suffix]).length, 0);
+    if (options.sequenceN) {
+      buffer.writeUInt32BE(options.sequenceN, 4);
+    }
 
-    // Concat buffers
-    const concatBuffer = Buffer.concat([prefix, len, payload, crc32Buffer, suffix]);
+    // Add payload, crc, and suffix
+    payload.copy(buffer, 16);
+    buffer.writeInt32BE(crc(buffer.slice(0, payload.length + 16)), payload.length + 16);
+    buffer.writeUInt32BE(0x0000AA55, payload.length + 20);
 
-    return concatBuffer;
+    return buffer;
   }
-}
 
-/**
-* Static wrapper for lower-level MessageParser
-* functions to easily parse packets.
-* @param {Buffer} data packet to parse
-* @returns {Object} result
-* @returns {String|Buffer|Object} result.data decoded data, if available in response
-* @returns {Number} result.commandByte command byte from decoded data
-*/
-function parse(data) {
-  const p = new MessageParser();
-  p._append(data);
-  p._parse();
-  return p._decode();
-}
+  /**
+   * Given a complete packet, writes
+   * a sequence number into the packet.
+   * @param {Buffer} buffer packet
+   * @param {Number} n sequence number
+   * @returns {Buffer}
+   */
+  writeSequenceN(buffer, n) {
```

Comment:
Why is this separate? Should be able to encode a message with the complete set of parameters. Use case where you might not?

Seems inconsistent with the other parameters of the message. If we want to construct a message piecemeal we should be able to set each parameter. But my preference would be enforcing a pattern where the message isn't encoded until you have all the parameters.

Reply:
This was an artifact from when I was testing other stuff out, removed.